libnet: setupDNS: don't overwrite user-modified resolv.conf

zhangguanzhang opened the pull request:
- Fix https://github.com/moby/moby/issues/51490#issuecomment-3528689655

**- What I did**

When restarting containers with non-host networks, `sandbox.setupDNS()` now calls `resolvconf.UserModified()` to check if resolv.conf was manually modified, preventing unintended overwrites.

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
Do not overwrite user-modified `/etc/resolv.conf` across container restarts.
```

**- A picture of a cute animal (not mandatory but encouraged)**
![7318e7f0b7850db8a698a4159692717f_720](https://github.com/user-attachments/assets/31f6ac3d-a1ad-4484-b044-79270a4d34da)

zhangguanzhang committed:
libnet: setupDNS: don't overwrite user-modified resolv.conf

Call resolvconf.UserModified() in sandbox.setupDNS() to check if
resolv.conf was manually modified before regenerating it during
container restart for non-host network modes.

Signed-off-by: zhangguanzhang <zhangguanzhang@qq.com>
Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>

zhangguanzhang commented:
Can someone help me take a look?

akerouanton commented on daemon/libnetwork/sandbox_dns_unix_test.go:
```diff
@@ -90,3 +95,69 @@ func TestDNSOptions(t *testing.T) {
+
+	// Step 2: Simulate cri-dockerd modifying the resolv.conf for a Kubernetes pause container.
+	// This mimics the behavior where external tools (like cri-dockerd) customize DNS
+	// settings for K8s pods, which should be preserved during container restart/unpause.
```
I'd prefer not to mention that very specific use-case. To be honest, it still makes little sense to me that you're trying to work around some of the k8s primitives in such a way, but I'm not going to fight over that.

In general, I think this enhancement makes sense though and we got another  request for that a few days ago. See https://github.com/moby/moby/issues/51570.

Could you please rephrase that in more general terms? Something like 'Step 2: Simulate user manually overwritting the container's resolv.conf'?

This would avoid future maintainers raising an eyebrow at this comment wondering what's the real use case.

akerouanton commented on daemon/libnetwork/sandbox_dns_unix_test.go:
```diff
@@ -90,3 +95,69 @@ func TestDNSOptions(t *testing.T) {
+	modifiedContent := []byte(`nameserver 10.96.0.10
+search default.svc.cluster.local. svc.cluster.local. cluster.local.
+options ndots:5
+`)
```
For the same reason, could you remove anything k8s related from here? A plain `nameserver 1.1.1.1` would do the job.

akerouanton committed:
libnet: populateNetworkResourcesOS: updateDNS only if !needResolver

When ep.needResolver() is true, sb.startResolver() calls sb.rebuildDNS()
which doesn't update the resolv.conf hash file.

Subsequent calls to sb.updateDNS() (which is only called by
populateNetworkResourcesOS) won't have any effect since it'll compare
the hash file and consider that the file was manually modified.

Make this explicit by gating the call to updateDNS() on !needResolver().

Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>

akerouanton committed:
libnet: rebuildDNS: update the hash file

Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>

akerouanton reviewed (approved):
Last commit LGTM.

robmry commented on daemon/libnetwork/sandbox_dns_unix_test.go:
```diff
@@ -90,3 +95,56 @@ func TestDNSOptions(t *testing.T) {
+	assert.NilError(t, err)
+
+	// Step 1: Create initial sandbox (simulating first container start)
+	sb, err := c.NewSandbox(context.Background(), "cnt1")
```
If the test fails and exits before getting to "step 3", `sb` needs to be deleted.

robmry commented on integration/networking/resolvconf_test.go:
```diff
@@ -212,3 +212,47 @@ func TestNslookupWindows(t *testing.T) {
+	)
+	defer c.ContainerRemove(ctx, id, client.ContainerRemoveOptions{
+		Force: true,
+	})
```
nit: busybox/top are set as defaults somewhere deep in the test client, so this can be just ...

```
	id := container.Run(ctx, t, c, container.WithName(ctrName))
	defer c.ContainerRemove(ctx, id, client.ContainerRemoveOptions{Force: true})
```

robmry reviewed (commented):
Thank you - just a couple of comments on the tests ...
